File: python/layer.py
import numpy as np
import math
import copy

class Layer(object):
    def __init__(self, input_shape = None, units = None):
        self.input_shape = input_shape
        self.units = units

    def configure(self, shape, phase, prevLayer):
        self.phase = phase
        self.prevLayer = prevLayer

    def forward(self, x):
        pass

    def backward(self, err_delta):
        pass


class Conv2D(Layer):
    def __init__(self, filters, kernel_size, strides=(1,1), input_shape = None, padding = 0, dtype=np.float32):
        super(Conv2D, self).__init__(input_shape, None)
        self.filterNum = filters
        self.kernel_size = kernel_size
        self.padding = padding
        self.filters = np.random.uniform(-1, 1, (filters, kernel_size**2)).astype(dtype)
        self.dtype = dtype
        self.strides = strides
        if input_shape:
            self.input_shape = (int(math.sqrt(input_shape))+self.padding)**2
            self.units = int(np.sqrt(self.input_shape)) - self.kernel_size + 1
            self.units *= self.units
            self.x_rowcol = int(np.sqrt(self.input_shape))
            self.y_rowcol = self.x_rowcol - self.kernel_size+1

    def configure(self, data_shape, phase, prevLayer=None):
        self.phase = phase
        self.prevLayer = prevLayer
        self.channel = data_shape[1]
        self.batch = data_shape[0]
        if prevLayer:
            self.units = int(np.sqrt(self.input_shape)) - self.kernel_size + 1
            self.units *= self.units
            self.input_shape = (int(math.sqrt(prevLayer.units)) + self.padding)**2
        self.x_rowcol = int(np.sqrt(self.input_shape))
        self.y_rowcol = (self.x_rowcol + 2 * self.padding - self.kernel_size)/self.strides[0] + 1
        self.X = np.zeros((self.batch, self.channel, self.kernel_size**2, self.y_rowcol**2), dtype=self.dtype)
        self.Y = np.zeros((self.batch, self.filterNum, self.y_rowcol**2), dtype=self.dtype)
        if phase == "TRAIN":
            self.E = np.zeros((self.batch, self.filterNum, self.y_rowcol**2), dtype=self.dtype)
            self.err_delta = np.zeros((self.batch, self.channel, self.x_rowcol, self.x_rowcol), dtype=self.dtype)

    def forward(self, x):
        self.Y = np.zeros((self.batch, self.filterNum, self.y_rowcol**2), dtype=self.dtype)
        x = x.reshape(self.batch, self.channel, self.x_rowcol-self.padding, self.x_rowcol-self.padding)
        if self.padding:
            x = np.lib.pad(x, (1,1), 'constant', constant_values=(0,0))

        for i in range(0, self.y_rowcol, self.strides[0]):
            for j in range(0, self.y_rowcol, self.strides[1]):
                self.X[:, :, :, self.y_rowcol*i+j%self.y_rowcol] = \
                x[:, :, i:i+self.kernel_size, j:j+self.kernel_size].reshape(self.batch, self.channel, self.kernel_size**2)

        for b in range(self.batch):
            for c in range(self.channel):
                self.Y[b, :] += self.filters.dot(self.X[b,c])
        return self.Y.reshape(self.batch, self.filterNum, self.y_rowcol, self.y_rowcol)

    def backward(self, err_delta):
        self.E[:] = err_delta.reshape(self.batch, self.filterNum, self.y_rowcol**2)

        for yi in range(self.y_rowcol):
            for yj in range(self.y_rowcol):
                # not good for vector processing
                tmp = np.einsum("bf,fk->bfk", self.E[:,:,yi*self.y_rowcol + yj%self.y_rowcol], self.filters)
                for f in range(self.filterNum):
                    yyi = yi * self.strides[0]
                    yyj = yj * self.strides[1]
                    self.err_delta[:, :, yyi:yyi+self.kernel_size, yyj:yyj+self.kernel_size] += \
                        tmp[:,f,:].reshape(self.batch, 1, self.kernel_size, self.kernel_size)

        tmp = np.zeros(self.filters.shape)
        for b in range(self.batch):
            for c in range(self.channel):
                tmp += self.X[b,c,:,:].dot(self.E[b,:,:].T).T
        self.filters -= tmp/self.batch

        return self.err_delta

class MaxPooling2D(Layer):
    def __init__(self, kernel_size, strides=(1,1), dtype=np.float32):
        super(MaxPooling2D, self).__init__()
        self.kernel_size = kernel_size
        self.strides = strides
        self.dtype = dtype

    def configure(self, data_shape, phase, prevLayer = None):
        self.phase = phase
        self.prevLayer = prevLayer
        # TODO : not good,
        self.x_rowcol = prevLayer.prevLayer.y_rowcol
        self.input_shape = self.x_rowcol**2
        self.y_rowcol = self.x_rowcol - self.kernel_size+1
        self.channel = prevLayer.prevLayer.filterNum
        self.batch = data_shape[0]
        self.X = np.zeros((self.batch, self.channel, self.kernel_size**2, self.y_rowcol**2), dtype=self.dtype)
        if phase == "TRAIN":
            self.maxLocs = np.zeros((self.batch, self.channel, self.y_rowcol**2), dtype=int)
            self.E = np.zeros((self.batch, self.channel, self.y_rowcol**2), dtype=self.dtype)
            self.err_delta = np.zeros((self.batch, self.channel, self.x_rowcol, self.x_rowcol), dtype=self.dtype)
            self.Y = np.zeros((self.batch, self.channel, self.y_rowcol**2), dtype=self.dtype)

    def forward(self, x):
        for i in range(0, self.y_rowcol, self.strides[0]):
            for j in range(0, self.y_rowcol, self.strides[1]):
                self.X[:, :, :, self.y_rowcol*i+j%self.y_rowcol] = \
                x[:, :, i:i+self.kernel_size, j:j+self.kernel_size].reshape(self.batch, self.channel, self.kernel_size**2)

        if self.phase == "TRAIN":
            self.maxLocs[:] = self.X.argmax(axis=2)
        return self.X.max(axis=2).reshape(self.batch, self.channel, self.y_rowcol, self.y_rowcol)
        #return self.Y.reshape(self.batch, self.channel, self.y_rowcol, self.y_rowcol)

    def backward(self, err_delta):
        self.E[:] = err_delta.reshape(self.batch, self.channel, self.y_rowcol**2)

        for b in range(self.batch):
            for c in range(self.channel):
                for y in range(self.y_rowcol**2):
                    # TODO : really bad way
                    loc = self.maxLocs[b,c,y]
                    locy = loc / self.kernel_size + (y / self.y_rowcol)
                    locx = loc % self.kernel_size + (y % self.y_rowcol)
                    self.err_delta[b,c,locx,locy] += self.E[b,c,y]

        return self.err_delta


class FullyConnect(Layer):
    def __init__(self, units, input_shape=0, dtype=np.float32):
        super(FullyConnect, self).__init__(input_shape, units)
        self.W = np.random.uniform(-1, 1, (input_shape, units)).astype(dtype)
        # TODO : sharing bias to all batch
        self.bias = np.random.uniform(-1, 1, 1).astype(dtype)
        self.original_shape = None
        self.dtype = dtype

    def configure(self, data_shape, phase, prevLayer = None):
        self.phase = phase
        self.prevLayer = prevLayer
        self.batch = data_shape[0]
        if isinstance(prevLayer, MaxPooling2D):
            self.input_shape = prevLayer.channel * prevLayer.y_rowcol * prevLayer.y_rowcol
        if phase == "TRAIN":
            self.X = np.zeros((self.batch, self.input_shape), dtype=self.dtype)
            self.E = np.zeros((self.batch, self.units), dtype=self.dtype)
            self.Y = np.zeros((self.batch, self.units), dtype=self.dtype)

    def forward(self, x):
        # for 2D data (like image)
        if len(x.shape) > 1:
            self.original_shape = x.shape
            x = x.reshape(self.original_shape[0], reduce(lambda x,y:x*y, self.original_shape[1:]))

        if self.phase == "TRAIN":
            self.X[:] = x
        return x.dot(self.W) + self.bias

    def backward(self, err_delta):
        self.E[:] = err_delta
        err_delta = self.E.dot(self.W.T)

        # updates
        np.subtract(self.W, self.optimizer(np.sum(np.einsum("bi,bj->bij", self.X, self.learning_rate*self.E), axis=0))/self.batch, self.W)
        self.bias -= np.sum(self.learning_rate * self.E)/self.batch

        if self.original_shape:
            err_delta = err_delta.reshape(self.original_shape)
        return err_delta

from activation import Sigmoid, Tanh
class LSTM(Layer):
    def __init__(self, units=0, input_shape=0, dtype=np.float32, gate_act=Sigmoid):
        super(LSTM, self).__init__(input_shape, units)
        self.gate_act = {"I":gate_act(), "F":gate_act(), "O":gate_act()}
        self.tanh = {"U":Tanh(), "C":Tanh()}
        self.Wx = {"I":None, "F":None, "U":None, "O":None}
        self.Wh = {"I":None, "F":None, "U":None, "O":None}
        self.B = {"I":None, "F":None, "U":None, "O":None}
        for k in ["I", "F", "U", "O"]:
            self.Wx[k] = np.random.uniform(-1, 1, (input_shape, units)).astype(dtype)
            self.Wh[k] = np.random.uniform(-1, 1, (units, units)).astype(dtype)
            self.B[k] = np.random.uniform(-1, 1, 1).astype(dtype)

    def configure(self, data_shape, phase, prevLayer = None):
        self.batch = data_shape[0]
        for k in self.gate_act:
            self.gate_act[k].configure(data_shape, phase, prevLayer)
        for k in self.tanh:
            self.tanh[k].configure(data_shape, phase, prevLayer)
        self.optimizers = []
        for i in range(8):
            self.optimizers.append(copy.deepcopy(self.optimizer))
        self.buff = {"C":None, "C_1":None, "H":None, "H_1":None, "I":None, "F":None, "U":None, "O":None, "X":None}
        for k in self.buff:
            self.buff[k] = np.zeros((self.batch, self.units))
        self.X = np.zeros((self.batch, self.input_shape), dtype=self.dtype)

    def forward(self, x):
        self.X[:] = x

        for k in ["I", "F", "O"]:
            self.buff[k] = self.gate_act[k].forward(self.X.dot(self.Wx[k]) + self.buff["H_1"].dot(self.Wh[k]) + self.B[k])
        self.buff["U"] = self.tanh["U"].forward(self.X.dot(self.Wx["U"]) + self.buff["H_1"].dot(self.Wh["U"]) + self.B["U"])
        self.buff["C"] = self.buff["I"]*self.buff["C_1"] + self.buff["U"] * self.buff["I"]
        self.Ctanh = self.tanh["C"].forward(self.buff["C"])
        self.buff["H"] = self.Ctanh * self.buff["O"]
        self.buff["C_1"] = self.buff["C"]
        self.buff["H_1"] = self.buff["H"]
        return self.buff["H"]

    def backward(self, e):
        delta = {}
        delta["C"] = self.tanh["C"].backward(e)*self.buff["O"]
        delta["C_1"] = delta["C"] * self.buff["F"]
        delta["O"] = self.gate_act["O"].backward(e)*self.Ctanh
        delta["I"] = self.gate_act["I"].backward(delta["C"])*self.buff["U"]
        delta["U"] = self.tanh["U"].backward(delta["C"])*self.buff["I"]
        delta["F"] = self.gate_act["F"].backward(delta["C"])*self.buff["C_1"]
        delta["H"] = delta["I"].dot(self.Wh["I"].T) + delta["O"].dot(self.Wh["O"].T) + delta["U"].dot(self.Wh["U"].T) + delta["F"].dot(self.Wh["F"].T)

        #update
        for i, k in enumerate(self.Wx):
            np.subtract(self.Wx[k], self.optimizers[i](np.sum(np.einsum("bi,bj->bij", self.X, self.learning_rate*delta[k]), axis=0))/self.batch, self.Wx[k])

        for i, k in enumerate(self.Wh):
            np.subtract(self.Wh["I"], self.optimizers[4+i](np.sum(np.einsum("bi,bj->bij", self.buff["H_1"], self.learning_rate*delta["I"]), axis=0))/self.batch, self.Wh["I"])

        for k in self.B:
            self.B[k] -= np.sum(self.learning_rate * delta[k])

        return delta["H"]


class BatchNorm(Layer):
    def __init__(self, units=0, input_shape=0, dtype=np.float32):
        super(BatchNorm, self).__init__(input_shape, units)
        self.beta = None
        self.gamma = None
        self.sub = None
        self.sq = None
        self.invsq = None
        self.var = None
        self.xhat = None
        self.epsilon = 10e-12

    def configure(self, data_shape, phase, prevLayer = None):
        self.phase = phase
        self.prevLayer = prevLayer
        self.input_shape = prevLayer.input_shape
        self.units = prevLayer.units
        self.batch = data_shape[0]
        self.beta = 0
        self.gamma = 1

    def forward(self, x):
        mean = np.sum(x, axis=0)/self.batch
        self.sub = x - mean
        self.var = np.var(x, axis=0)
        self.sq = np.sqrt(self.var + self.epsilon)
        self.invsq = 1/self.sq
        self.xhat = self.sub*self.invsq
        return self.gamma * self.xhat + self.beta

    def backward(self, e):
        dxhat = e * self.gamma
        dvar = np.sum(dxhat * self.sub * self.sq**(1.5)/-2, axis=0)
        dmean = np.sum(dxhat * - self.invsq, axis=0) + dvar * -2 * self.var
        dx = dxhat * self.invsq + dvar * -2 * self.var + dmean / self.batch
        db = np.sum(e, axis=0)
        dg = db * self.xhat
        self.beta -= db
        self.gamma -= dg
        return dx

File: cpp/layer.cc
#include "layer.h"
#include <stdlib.h>
#include <random>

Layer::Layer(int input_shape, int units) : batch(1), filter(0), channel(0), input_shape(input_shape), units(units), prevLayer(nullptr), momentum_a(1.0) {}
Layer::~Layer() {
    this->Y.clear();
    this->Y.shrink_to_fit();
    if (this->phase == TRAIN) {
        this->E.clear();
        this->E.shrink_to_fit();
    }
}

int Layer::configure(int batch, float learning_rate, float v_param, Layer* prevLayer, phase_t phase) {
    this->batch = batch;
    this->batch_inv = 1/batch;
    this->learning_rate = learning_rate;
    this->momentum_a = v_param;
    this->phase = phase;
    std::random_device rd;
    this->mt = std::mt19937(rd());
    this->rand = std::uniform_real_distribution<float>(-1.0,1.0);
    if (prevLayer != nullptr) {
        prevLayer->nxtLayer = this;
        this->prevLayer = prevLayer;
    }
    return 1;
}

FullyConnect::FullyConnect(int input_shape, int units) : Layer(input_shape, units) {}
FullyConnect::~FullyConnect() {
    this->W.clear();
    this->W.shrink_to_fit();
    if (this->phase == TRAIN) {
        this->delta_buf.clear();
        this->delta_buf.shrink_to_fit();
    }
}

int FullyConnect::configure(int batch, float learning_rate, float v_param, Layer* prevLayer, phase_t phase) {
    Layer::configure(batch, learning_rate, v_param, prevLayer, phase);
    this->Y.resize(this->batch*this->units);
    this->W.resize(this->input_shape*this->units);
    if (this->phase == TRAIN) {
        this->E.resize(this->batch*this->input_shape);
        this->delta_buf.resize(this->batch*this->units);
    }

    for (int b = 32; b > 0; b /= 2) {
        if (this->batch % b == 0) {
            this->blkB = b;
            break;
        }
    }
    for (int i = 32; i > 0; i /= 2) {
        if (this->input_shape % i == 0) {
            this->blkI = i;
            break;
        }
    }
    for (int u = 32; u > 0; u /= 2) {
        if (this->input_shape % u == 0) {
            this->blkU = u;
            break;
        }
    }
    #pragma omp  parallel for
    for (int iu = 0; iu < this->input_shape*this->units; iu++) {
            this->W[iu] = this->rand(this->mt);
    }
    this->B = rand(mt);
    return 1;
}

void FullyConnect::forward(vector<float> *x) {
    #pragma omp  parallel for
    for (int i = 0; i < this->batch*this->units; i++) {
        this->Y[i] = 0;
    }
    if (this->blkB > 0 && this->blkI > 0 && this->blkU > 0) {
        #pragma omp parallel for
        for (int bb = 0; bb < this->batch; bb += this->blkB) {
            for (int ib = 0; ib < this->input_shape; ib += this->blkI) {
                for (int ub = 0; ub < this->units; ub += this->blkU) {
                    for (int b = 0; b < bb+this->blkB; b++) {
                        for (int i = 0; i < ib+this->blkI; i++) {
                            for (int u = 0; u < ub+this->blkU; u++) {
                                this->Y[b*this->units + u] += x->at(b*this->input_shape + i) * this->W[i*this->units + u] + this->B;
                        }}}
            }}}
    } else {
        #pragma omp  parallel for
        for (int b = 0; b < this->batch; b++) {
            for (int i = 0; i < this->input_shape; i++) {
                for (int u = 0; u < this->units; u++) {
                    this->Y[b*this->units + u] += x->at(b*this->input_shape + i) * this->W[i*this->units + u] + this->B;
                }
            }
        }
    }
    if (this->phase == TRAIN) {
        this->X = x;
    }
    return;
}

void FullyConnect::backward(vector<float> *e) {
    #pragma omp  parallel for
    for (int i = 0; i < this->batch*this->input_shape; i++) {
        this->E[i] = 0;
    }
    #pragma omp  parallel for
    for (int b = 0; b < this->batch; b++) {
        for (int u = 0; u < this->units; u++) {
            for (int i = 0; i < this->input_shape; i++) {
                this->E[b*this->input_shape + i] += e->at(b*this->units + u) * this->W[i*this->units+u];
            }
        }
    }

    // update
    if (this->momentum_a != 0) {
        #pragma omp  parallel for
        for (int b = 0; b < this->batch; b++) {
            for (int i = 0; i < this->input_shape; i++) {
                for (int u = 0; u < this->units; u++) {
                    this->W[i*this->units + u] -= this->momentum_a * this->delta_buf[b*this->units + u] + \
                        (this->learning_rate * this->X->at(b*this->input_shape + i) * e->at(b*this->units + u))*this->batch_inv;
                }
            }
        }
        #pragma omp parallel for
        for (int i = 0; i < this->batch*this->units; i++) {
            this->delta_buf[i] = e->at(i);
        }
    } else {
        for (int b = 0; b < this->batch; b++) {
            for (int i = 0; i < this->input_shape; i++) {
                for (int u = 0; u < this->units; u++) {
                    this->W[i*this->units + u] -= this->learning_rate * this->X->at(b*this->input_shape + i) * e->at(b*this->units + u)*this->batch_inv;
                }
            }
        }
    }

    #pragma omp  parallel for
    for (int b = 0; b < this->batch; b++) {
        for (int u = 0; u < this->units; u++) {
            this->B -= this->learning_rate * e->at(b*this->units + u)*this->batch_inv;
        }
    }

    return;
}


Conv2D::Conv2D(int input_shape, int channel, int filter, int kernel_size, int stride, int padding) : Layer(input_shape, 0), i_rowcol((int)std::sqrt((float)input_shape)), kernel_size(kernel_size), stride(stride), padding(padding) {
    if (stride <= 0) {
        // warning
        this->stride = 1;
    }
    this->u_rowcol = (this->i_rowcol + 2*padding - kernel_size)/stride + 1;
    this->units = this->u_rowcol*this->u_rowcol;
    this->channel = channel;
    this->filter = filter;
}
Conv2D::~Conv2D() {
    this->F.clear();
    this->F.shrink_to_fit();
    if (this->phase == TRAIN) {
        this->delta_buf.clear();
        this->delta_buf.shrink_to_fit();
    }
}

int Conv2D::configure(int batch, float learning_rate, float v_param, Layer* prevLayer, phase_t phase) {
    Layer::configure(batch, learning_rate, v_param, prevLayer, phase);
    // for filter and data matmul
    //this->X = (float*)malloc(sizeof(float)*this->batch*this->channel*this->kernel_size*this->kernel_size*this->units*this->units);
    this->Y.resize(this->batch*this->filter*this->units);
    this->E.resize(this->batch*this->channel*this->input_shape);
    this->delta_buf.resize(this->batch*this->filter*this->units);
    this->F.resize(this->filter*this->kernel_size*this->kernel_size);

    #pragma omp  parallel for
    for (int i = 0; i < this->filter*this->kernel_size*this->kernel_size; i++) {
        this->F[i] = this->rand(this->mt);
    }

    return 1;
}

void Conv2D::forward(vector<float> *x) {
    #pragma omp  parallel for
    for (int i = 0; i < this->batch*this->filter*this->units; i++) {
        this->Y[i] = 0;
    }
    #pragma omp  parallel for
    for (int b = 0; b < this->batch; b++) {
        for (int c = 0; c < this->channel; c++) {
            for (int f = 0; f < this->filter; f++) {
                for (int ro = 0; ro + this->kernel_size < this->i_rowcol; ro += this->stride) {
                    for (int co = 0; co + this->kernel_size < this->i_rowcol; co += this->stride) {
                        for (int ki = 0; ki < this->kernel_size; ki++) {
                            for (int kj = 0; kj < this->kernel_size; kj++) {
                                this->Y[b*this->units*this->filter+f*this->units+ro*this->i_rowcol+co] += \
                                    x->at(b*this->input_shape*this->channel+c*this->input_shape+ \
                                          ro*this->i_rowcol+co+ki*this->i_rowcol+kj)* \
                                    this->F[f*this->kernel_size*this->kernel_size+ki*this->kernel_size+kj];
                            }
                        }
                    }
                }
            }
        }
    }

    if (this->phase == TRAIN) {
        this->X = x;
    }
    return;
}


void Conv2D::backward(vector<float> *e) {
    #pragma omp  parallel for
    for (int i = 0; i < this->batch*this->channel*this->input_shape; i++) {
        this->E[i] = 0;
    }
    #pragma omp  parallel for
    for (int b = 0; b < this->batch; b++) {
        for (int c = 0; c < this->channel; c++) {
            for (int f = 0; f < this->filter; f++) {
                for (int ro = 0; ro < this->u_rowcol; ro++) {
                    for (int co = 0; co < this->u_rowcol; co++) {
                        for (int ki = 0; ki < this->kernel_size; ki++) {
                            for (int kj = 0; kj < this->kernel_size; kj++) {
                                this->E[b*this->channel*this->input_shape+c*this->input_shape + \
                                        ro*this->stride+co*this->stride+ki*this->i_rowcol+kj] += \
                                    this->Y[b*this->filter*this->units+f*this->units+ro*this->u_rowcol+co] * \
                                    this->F[f*this->kernel_size*this->kernel_size+ki*this->kernel_size+kj];
                            }
                        }
                    }
                }
            }
        }
    }

    // update
    if (this->momentum_a != 0) {
        #pragma omp  parallel for
        for (int b = 0; b < this->batch; b++) {
            for (int c = 0; c < this->channel; c++) {
                for (int f = 0; f < this->filter; f++) {
                    for (int ro = 0; ro < this->u_rowcol; ro++) {
                        for (int co = 0; co < this->u_rowcol; co++) {
                            for (int ki = 0; ki < this->kernel_size; ki++) {
                                for (int kj = 0; kj < this->kernel_size; kj++) {
                                    this->F[f*this->kernel_size*this->kernel_size+ki*this->kernel_size+kj] -= \
                                        this->momentum_a * this->delta_buf[b*this->filter*this->units+f*this->units+ro*this->u_rowcol+co] + \
                                        (this->learning_rate * this->X->at(b*this->channel*this->input_shape+c*this->input_shape+ \
                                                                           ro*this->i_rowcol+co+ki*this->i_rowcol+kj) * \
                                         e->at(b*this->filter*this->units+f*this->units+ro*this->u_rowcol+co))*this->batch_inv;
                                }
                            }
                        }
                    }
                }
            }
        }
        #pragma omp parallel for
        for (int i = 0; i < this->batch*this->filter*this->units; i++) {
            this->delta_buf[i] = e->at(i);
        }
    } else {
        #pragma omp  parallel for
        for (int b = 0; b < this->batch; b++) {
            for (int c = 0; c < this->channel; c++) {
                for (int f = 0; f < this->filter; f++) {
                    for (int ro = 0; ro < this->u_rowcol; ro++) {
                        for (int co = 0; co < this->u_rowcol; co++) {
                            for (int ki = 0; ki < this->kernel_size; ki++) {
                                for (int kj = 0; kj < this->kernel_size; kj++) {
                                    this->F[f*this->kernel_size*this->kernel_size+ki*this->kernel_size+kj] -= \
                                        this->learning_rate * this->X->at(b*this->channel*this->input_shape+c*this->input_shape+ \
                                                                          ro*this->i_rowcol+co+ki*this->i_rowcol+kj) * \
                                        e->at(b*this->filter*this->units+f*this->units+ro*this->u_rowcol+co)*this->batch_inv;
                                }
                            }
                        }
                    }
                }
            }
        }
    }


    return;
}


MaxPooling2D::MaxPooling2D(int input_shape, int channel, int kernel_size, int stride) : Layer(input_shape, 0), i_rowcol((int)std::sqrt((float)input_shape)), kernel_size(kernel_size), stride(stride) {
    if (stride <= 0) {
        // warning
        this->stride = 1;
    }
    this->u_rowcol = (this->i_rowcol - kernel_size)/stride + 1;
    this->units = this->u_rowcol*this->u_rowcol;
    this->channel = channel;
}
MaxPooling2D::~MaxPooling2D() {
    if (this->phase == TRAIN) {
        this->L.clear();
        this->L.shrink_to_fit();
    }
}

int MaxPooling2D::configure(int batch, float learning_rate, float v_param, Layer* prevLayer, phase_t phase) {
    Layer::configure(batch, learning_rate, v_param, prevLayer, phase);
    this->Y.resize(this->batch*this->channel*this->units);
    if (this->phase == TRAIN) {
        this->L.resize(this->batch*this->channel*this->units);
        this->E.resize(this->batch*this->channel*this->input_shape);
    }
    return 1;
}


void MaxPooling2D::forward(vector<float>* x) {
    float tmp;
    #pragma omp  parallel for private(tmp)
    for (int b = 0; b < this->batch; b++) {
        for (int c = 0; c < this->channel; c++) {
            for (int ro = 0; ro < this->u_rowcol; ro++) {
                for (int co = 0; co < this->u_rowcol; co++) {
                    tmp = x->at(b*this->channel*this->input_shape+c*this->input_shape+ro*this->i_rowcol+co);
                    for (int ki = 0; ki < this->kernel_size; ki++) {
                        for (int kj = 1; kj < this->kernel_size; kj++) {
                            if (tmp < x->at(b*this->channel*this->input_shape+c*this->input_shape+ro*this->i_rowcol+co+ki*this->i_rowcol+kj)) {
                                tmp = x->at(b*this->channel*this->input_shape+c*this->input_shape+ro*this->i_rowcol+co+ki*this->i_rowcol+kj);
                                this->Y[b*this->channel*this->units+c*this->units+ro*this->u_rowcol+co] = tmp;
                                if (this->phase == TRAIN) {
                                    this->L[b*this->channel*this->units+c*this->units+ro*this->u_rowcol+co] = ki*this->i_rowcol+kj;
                                }
                            }
                        }
                    }
                }
            }
        }
    }
    return;
}


void MaxPooling2D::backward(vector<float> *e) {
    #pragma omp  parallel for
    for (int i = 0; i < this->batch*this->channel*this->input_shape; i++) {
        this->E[i] = 0;
    }
    #pragma omp  parallel for
    for (int b = 0; b < this->batch; b++) {
        for (int c = 0; c < this->channel; c++) {
            for (int ro = 0; ro < this->u_rowcol; ro++) {
                for (int co = 0; co < this->u_rowcol; co++) {
                    this->E[b*this->input_shape*this->channel+c*this->input_shape+ro*this->i_rowcol+co+ \
                            this->L[b*this->units*this->channel+c*this->units+ro*this->u_rowcol+co]] = \
                        e->at(b*this->units*this->channel+c*this->units+ro*this->u_rowcol+co);
                }
            }
        }
    }
    return;
}

/*
LSTM::LSTM(int input_shape, int units) : Layer(input_shape, units) {}
~LSTM::LSTM() {
    for (int i = 0; i < 4; i++) {
        free(this->Wx[this->Idx[i]]);
        free(this->Wh[this->Idx[i]]);
    }
}

int LSTM::configure(int batch, float learning_rate, float v_param, Layer* prevLayer, phase_t phase) {
    int ret = Layer::configure(batch, learning_rate, v_param, prevLayer, phase);
    this->acts['I'] = new Sigmoid();
    this->acts['F'] = new Sigmoid();
    this->acts['O'] = new Sigmoid();
    this->acts['U'] = new Tanh();

    std::random_device rd;
    std::mt19937 mt(rd());
    std::uniform_real_distribution<float> rand(-1.0,1.0);
    for (int i = 0; i < 4; i++) {
        this->Wx[this->Idx[i]] = (float*)malloc(sizeof(float*)*this->input_shape*this->units);
        this->Wx[this->Idx[i]] = (float*)malloc(sizeof(float*)*this->units*this->units);
        #pragma omp  parallel for
        for (int iu = 0; iu < this->input_shape*this->units; iu++) {
            this->Wx[this->Idx[i]][iu] = this->rand(this->mt);
        }
        #pragma omp  parallel for
        for (int uu = 0; uu < this->units*this->units; uu++) {
            this->Wh[this->Idx[i]][uu] = this->rand(this->mt);
        }
        *this->B[this->Idx[i]] = this->rand(this->mt);
        this->acts[this->Idx].configure(batch, learning_rate, v_param, prevLayer, phase);
    }

    for (int i = 0; i < 8; i++) {
        this->buff[this->bIdx[i]] = (float*)malloc(sizeof(float*)*this->batch*this->units);
    }

    return ret;
}

void LSTM::forward(float* x) {
    for (int g = 0; i < 4; i++) {
        for (int b = 0; b < this->batch; b++) {
            for (int i = 0; i < this->input_shape; i++) {
                for (int u = 0; u < this->units; u++) {
                    this->buff[this->Idx[g]][b*this->units + u] += x[b*this->input_shape + i] * this->Wx[this->Idx[g]][i*this->units + u] + this->buff['h'][b*this->units + u] * this->Wh[this->Idx[g]][u*this->units + u] + *this->B;
                }
            }
        }
    }

    if (this->phase = TRAIN) {
        this->X = x;
    }
    this->Y = this->buff["H"];
}
*/
